crates/chain/src/local_chain.rs

```diff
@@ -17,224 +119,450 @@ impl ChainOracle for LocalChain {
     fn is_block_in_chain(
         &self,
         block: BlockId,
-        static_block: BlockId,
+        chain_tip: BlockId,
     ) -> Result<Option<bool>, Self::Error> {
-        if block.height > static_block.height {
+        if block.height > chain_tip.height {
             return Ok(None);
         }
         Ok(
             match (
-                self.blocks.get(&block.height),
-                self.blocks.get(&static_block.height),
+                self.checkpoints.get(&block.height),
+                self.checkpoints.get(&chain_tip.height),
             ) {
-                (Some(&hash), Some(&static_hash)) => {
-                    Some(hash == block.hash && static_hash == static_block.hash)
+                (Some(cp), Some(tip_cp)) => {
+                    Some(cp.hash == block.hash && tip_cp.hash == chain_tip.hash)
                 }
                 _ => None,
             },
         )
     }
-}
 
-impl AsRef<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn as_ref(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    fn get_chain_tip(&self) -> Result<Option<BlockId>, Self::Error> {
+        Ok(self.checkpoints.values().last().map(|cp| BlockId {
+            height: cp.height,
+            hash: cp.hash,
+        }))
     }
 }
 
-impl From<LocalChain> for BTreeMap<u32, BlockHash> {
-    fn from(value: LocalChain) -> Self {
-        value.blocks
-    }
-}
-
-impl From<BTreeMap<u32, BlockHash>> for LocalChain {
-    fn from(value: BTreeMap<u32, BlockHash>) -> Self {
-        Self { blocks: value }
+impl LocalChain {
+    /// Construct a [`LocalChain`] from an initial `changeset`.
+    pub fn from_changeset(changeset: ChangeSet) -> Self {
+        let mut chain = Self::default();
+        chain.apply_changeset(&changeset);
+        chain
     }
-}
 
-impl LocalChain {
-    pub fn from_blocks<B>(blocks: B) -> Self
-    where
-        B: IntoIterator<Item = BlockId>,
-    {
-        Self {
-            blocks: blocks.into_iter().map(|b| (b.height, b.hash)).collect(),
+    /// Construct a [`LocalChain`] from a given `checkpoint` tip.
+    pub fn from_checkpoint(checkpoint: Arc<CheckPoint>) -> Result<Self, InvalidHeightOrderError> {
+        match Self::default().extend_from_last_agreement(checkpoint) {
+            Ok((chain, _)) => Ok(chain),
+            Err(ExtendError::InvalidHeightOrder(err)) => Err(err),
+            Err(ExtendError::CannotConnect(_)) => {
+                unreachable!("we do not need to connect to an empty chain")
+            }
         }
     }
 
-    /// Get a reference to a map of block height to hash.
-    pub fn blocks(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    /// Constructs a [`LocalChain`] from an iterator of [`BlockId`]s.
+    ///
+    /// The blocks do not need to be in order. If multiple blocks of the same height are introduced,
+    /// the latest block will be used.
+    pub fn from_blocks(blocks: impl IntoIterator<Item = BlockId>) -> Self {
```

Comment:
Not sure about this API. Can't we just force them to pass in a `BTreeSet<BlockId>` so it's ordered properly and so we can set `prev` in a natural way.

The docs must say that it is up to the caller to make sure the blocks are in the same chain.

Reply:
Changed as suggested.